README.md
# NAME

Web::API::Mock - It's new $module

# SYNOPSIS

    $ git clone  git@github.com:takihito/Web-API-Mock.git 
    $ cpanm ./Web-API-Mock
    
    or
    
    $ cpanm Web::API::Mock
    
    :
    $ run-api-mock --help
    Usage:
            $ run-api-mock --files api.md --not-implemented-urls url.txt --port 8080

# DESCRIPTION

\## Install 

<pre>
$ git clone  git@github.com:takihito/Web-API-Mock.git 
$ cpanm ./Web-API-Mock

or

$ cpanm Web::API::Mock

:
$ run-api-mock --help 
Usage:
        $ run-api-mock --files api.md --not-implemented-urls url.txt --port 8080
</pre>

\## Create API Documentation

\* http://apiblueprint.org

<pre>
\# api.md 

\## GET /hello

\+ Response 200 (text/plain)

        Hello World
</pre>

\## Start Mock Server

<pre>
$ run-api-mock --files api.md 
HTTP::Server::PSGI: Accepting connections at http://0:5000
</pre>

\## Options

\* files
 \* API Document Files.

<pre>
$ run-api-mock --files api1.md --files api2.md
</pre>

\* not-implemented-urls
 \* Add "501 Not Implemented" info (Method and URL). 

<pre>
GET,/hello/hoge
POST,/hello/foo
</pre>

\* port
 \* server port

\* Other...
 \* plackup options 

\## Switch to the Production API 

By using nginx 

see not-implemented-urls option.

<pre>
$ run-api-mock --files api.md --not-implemented-urls url.txt --port 5001
</pre>

<pre>
upstream mock\_backend {
   server 127.0.0.1:5001;
}

upstream prod\_backend {
   server 127.0.0.1:5002;
}

server {

: 
: 

    location ~ ^/ {
         proxy_intercept_errors on;
         proxy_pass http://mock_backend;
         proxy_set_header Host $host;
    }

    error_page 501 =200 @prod;
    location @prod {
        proxy_pass http://prod_backend;
        proxy_set_header Host $host;
    }
</pre>

# LICENSE

Copyright (C) akihito.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

akihito <takeda.akihito@gmail.com>


Changes
Revision history for Perl extension Web-API-Mock

{{$NEXT}}

0.01 2015-02-25T10:40:10Z

    - original version



lib/Web/API/Mock/Map.pm
package Web::API::Mock::Map;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";

use  Web::API::Mock::Resource;

use Class::Accessor::Lite (
    new => 1,
    rw  => [ qw/resources/ ]
);

sub init {
    my ($self) = @_;

    $self->resources({});

    $self->add_resource( '/',  Web::API::Mock::Resource->status_404);
}


sub add_resource {
    my ($self, $url, $args) = @_;

    my $resource = $self->resources->{$url} || Web::API::Mock::Resource->new();
    $resource->add({
        status       => $args->{status},
        content_type => $args->{content_type},
        method       => $args->{method},
        header       => $args->{header},
        body         => $args->{body}
    });
    $self->resources->{$url} = $resource;
}

sub request {
    my ($self, $method, $url) = @_;

    my $resource = $self->resources->{$url} ? $self->resources->{$url} : '';
    unless ($resource) {
        $url =~ s!^(.+\/).+?$!$1\{\.+?}!;
        ($url) = grep { m!^$url! } @{$self->url_list};
        if ( $url && $self->resources->{$url} ) {
            $resource = $self->resources->{$url};
        }
    }
    return $resource->response($method) if $resource;

    return;
}

sub url_list {
    my ($self) = @_;
    return [keys %{$self->resources}];
}

1;


lib/Web/API/Mock/Parser.pm
package Web::API::Mock::Parser;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";

use utf8;
use B qw(perlstring);
use Text::Markdown::Hoedown;
use Web::API::Mock::Map;
use Class::Accessor::Lite (
    new => 1,
    rw  => [ qw/map md cb _resource_name _paragraph _url _method _status _content_type _header _body api/ ],
);

sub create_map {
    my $self =shift;

    $self->_callback();

    my $md = Text::Markdown::Hoedown::Markdown->new(HOEDOWN_EXT_AUTOLINK, 16, $self->cb);

    eval {
        $md->render($self->md);
    };
    if ( $@ ) {
        die "cannot parse markdown. $@";
    }

    my $map = Web::API::Mock::Map->new();
    $map->init();
    for my $name ( keys %{$self->api} ) {
        $map->add_resource( $self->api->{$name}->{url}, {
            status       => $self->api->{$name}->{status},
            content_type => $self->api->{$name}->{content_type},
            method       => $self->api->{$name}->{method},
            header       => $self->api->{$name}->{header},
            body         => $self->api->{$name}->{body}
        });
    }

    return $map;
}

sub _callback {
    my $self = shift;

    $self->api({});

    my $cb = Text::Markdown::Hoedown::Renderer::Callback->new();

    $cb->blockcode( sub {
        return unless $self->_resource_name;

        if ( $self->_paragraph =~ /^body/i ) {
           $self->_body($_[0]);
        }
        elsif ( $self->_paragraph =~ /^header/i ) {
            my @line = split("\n", $_[0]);
            for (@line) {
                my ($key, $value) = split(':',$_);
                chomp $value;
                $value =~ s/^ +//g;
                $self->_header->{$key} = $value;
            }
        }
        else {
           $self->_body($_[0]);
        }

    } );

    # TODO 
    $cb->blockquote( sub {} );

    # TODO
    $cb->blockhtml( sub {} );

    $cb->header( sub {
        $self->_add_api();

        my ($method, $url) = split(' ', $_[0]);
        $self->_header({});
        $self->_body('');
        $self->_url($url);
        $self->_method($method);
        $self->_resource_name($_[0]);
    } );

    $cb->paragraph( sub {
        $self->_paragraph($_[0]);
    } );

    # TODO 
    $cb->list(sub{});

    $cb->listitem(sub{
        if ( $_[0] =~ /^response +(\d+) +\((.+?)\)/i ) {
            $self->_status($1);
            $self->_content_type($2);
        }
    });

    $cb->doc_footer(sub{
        $self->_add_api();
    });

    $self->cb($cb);
}

sub _add_api {
    my $self = shift;
    if ( $self->_resource_name && $self->_url && $self->_method && $self->_method =~ /^(GET|POST|PUT|DELETE)$/ ){
        $self->api->{$self->_resource_name} = {
            content_type => $self->_content_type // '',
            status => $self->_status // '',
            url    => $self->_url // '',
            method => $self->_method // '',
            header => $self->_header // {},
            body   => $self->_body // '',
        };
    }
}

1;


lib/Web/API/Mock/Resource.pm
package Web::API::Mock::Resource;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";

use Class::Accessor::Lite (
    new => 1,
    rw  => [ qw/url header body status content_type/ ],
);

sub add {
    my ($self, $args) = @_;
    my $method = $args->{method} // 'GET';
    my $header = $args->{header} // {};
    my $body   = $args->{body} // '';
    my $status = $args->{status} // 200;
    my $content_type = $args->{content_type} // 'text/html';

    unless ( $self->body ) {
        $self->status({});
        $self->content_type({});
        $self->header({});
        $self->body({});
    }

    $self->status->{$method} = $status;
    $self->content_type->{$method} = $content_type;
    $self->header->{$method} = $header;
    $self->body->{$method} = $body;
}

sub response {
    my ($self, $method) = @_;
    $method //= 'GET';

    return {
        status       => $self->status->{$method},
        content_type => $self->content_type->{$method},
        header       => $self->header->{$method},
        body         => $self->body->{$method}
    }

}


1;


lib/Web/API/Mock.pm
package Web::API::Mock;
use 5.008005;
use strict;
use warnings;

use Plack::Request;
use Web::API::Mock::Parser;
use Web::API::Mock::Resource;
use Class::Accessor::Lite (
    new => 1,
    rw  => [ qw/config files not_implemented_urls map/ ],
);

our $VERSION = "0.01";

sub setup {
    my ($self, $files, $not_implemented_url_file) = @_;
    my $markdown;
    for my $file (@{$files}) {
        open my $fh, "<:encoding(utf8)", $file or die "cannot open file. $file:$!";
        while ( my $line = <$fh> ) {
            $markdown .= $line;
        }
        close($fh);
    }
    my $parser = Web::API::Mock::Parser->new();
    $parser->md($markdown);
    $self->map($parser->create_map());

    $self->not_implemented_urls([]);
    if ($not_implemented_url_file) {
        open my $fh, "<:encoding(utf8)", $not_implemented_url_file or die "cannot open file. $not_implemented_url_file:$!";
        while ( my $line = <$fh> ) {
            chomp $line;
            $line =~ s/\ //g;
            push @{$self->not_implemented_urls}, $line;
        }
        close($fh);
    }
}

sub psgi {
    my $self = shift;
    sub {
        my $env = shift;

        my $req = Plack::Request->new($env);
        my $plack_response = $req->new_response(404);
        my $response  = Web::API::Mock::Resource->status_404;

        if ($self->check_implemented_url($req->method, $req->path_info)) {
            $response  = Web::API::Mock::Resource->status_501;
        }
        else {
            $response = $self->map->request($req->method, $req->path_info);
            if (!$response || !$response->{status}) {
                $response  = Web::API::Mock::Resource->status_404;
            }
        }

        $plack_response->headers($response->{header});
        $plack_response->content_type($response->{content_type});
        $plack_response->status($response->{status});
        $plack_response->body($response->{body});
        $plack_response->finalize;
    };
}

sub check_implemented_url {
    my ($self, $method, $path) = @_;

    return if ( ref $self->not_implemented_urls ne 'ARRAY');

    my $target = join(',', $method,$path);
    my ($url) = grep { m!^$target$! } @{$self->not_implemented_urls};
    # TODO 再帰
    unless ($url) {
        $target =~ s!^(.+\/).+?$!$1\{.+?}!;
        ($url) = grep { m!^$target$! } @{$self->not_implemented_urls};
    }
    return $url;
}

1;
__END__

=encoding utf-8

=head1 NAME

Web::API::Mock - It's new $module

=head1 SYNOPSIS

    $ git clone  git@github.com:takihito/Web-API-Mock.git 
    $ cpanm ./Web-API-Mock
    
    or
    
    $ cpanm Web::API::Mock
    
    :
    $ run-api-mock --help
    Usage:
            $ run-api-mock --files api.md --not-implemented-urls url.txt --port 8080

=head1 DESCRIPTION

## Install 

<pre>
$ git clone  git@github.com:takihito/Web-API-Mock.git 
$ cpanm ./Web-API-Mock

or

$ cpanm Web::API::Mock

:
$ run-api-mock --help 
Usage:
        $ run-api-mock --files api.md --not-implemented-urls url.txt --port 8080
</pre>


## Create API Documentation

* http://apiblueprint.org

<pre>
# api.md 

## GET /hello

+ Response 200 (text/plain)

        Hello World
</pre>

## Start Mock Server

<pre>
$ run-api-mock --files api.md 
HTTP::Server::PSGI: Accepting connections at http://0:5000
</pre>

## Options

* files
 * API Document Files.

<pre>
$ run-api-mock --files api1.md --files api2.md
</pre>

* not-implemented-urls
 * Add "501 Not Implemented" info (Method and URL). 

<pre>
GET,/hello/hoge
POST,/hello/foo
</pre>

* port
 * server port

* Other...
 * plackup options 

## Switch to the Production API 

By using nginx 

see not-implemented-urls option.

<pre>
$ run-api-mock --files api.md --not-implemented-urls url.txt --port 5001
</pre>

<pre>
upstream mock_backend {
   server 127.0.0.1:5001;
}

upstream prod_backend {
   server 127.0.0.1:5002;
}

server {

: 
: 

    location ~ ^/ {
         proxy_intercept_errors on;
         proxy_pass http://mock_backend;
         proxy_set_header Host $host;
    }

    error_page 501 =200 @prod;
    location @prod {
        proxy_pass http://prod_backend;
        proxy_set_header Host $host;
    }
</pre>

=head1 LICENSE

Copyright (C) akihito.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

akihito E<lt>takeda.akihito@gmail.comE<gt>

=cut

